Flaky Network Server tests

<!--
Thanks for submitting a bug report. Please fill the template below,
otherwise we will not be able to process this bug report.
-->

#### Summary
<!-- Summarize the problem in a few sentences: -->

Running the Network Server unit tests fails.

#### Steps to Reproduce
<!-- How can we reproduce the problem? -->

1. `git checkout v3.11`
2. `docker run --rm -it -p 6379:6379 redis:5` 
3. `go test ./pkg/networkserver -failfast`

<!--
Please upload relevant configuration (as .txt).
If you use the command "ttn-lw-stack config", you can redact sensitive config.
-->

#### What do you see now?
<!-- Please paste terminal output, upload logs (as .txt) or upload screenshots. -->

```
$ go test ./pkg/networkserver/redis/ -failfast
--- FAIL: TestDownlinkTaskQueue (0.01s)
    --- FAIL: TestDownlinkTaskQueue/1st_run (0.00s)
        downlink_task_queue.go:95: Pop returned on empty schedule with error: 
            -> error:pkg/redis:store (store error) (attributes: map[])
            --> ERR Error running script (call to f_32bd83583b1367d88c7534ddb33d6622a3c58cc7): @user_script:8: @user_script: 8: This Redis command is not allowed from scripts  (attributes: map[])
    downlink_task_queue_test.go:32: Skipping 2nd run
--- FAIL: TestApplicationUplinkQueue (0.01s)
    --- FAIL: TestApplicationUplinkQueue/1st_run (0.00s)
        application_uplink_queue.go:198: Received unexpected Pop error: error:pkg/redis:store (store error)
        doc.go:66: 
            /home/ubuntu/lorawan-stack/pkg/networkserver/internal/test/shared/application_uplink_queue.go:198
            Expected: true
            Actual:   false
    application_uplink_queue_test.go:32: Skipping 2nd run
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 995ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
redis: 2021/01/11 23:39:16 commands.go:36: specified duration is 64ms, but minimal supported value is 1s - truncating to 1s
FAIL
FAIL	go.thethings.network/lorawan-stack/v3/pkg/networkserver/redis	0.270s
FAIL

```

CI also seems to fail randomly on a different test (see failure on https://github.com/TheThingsNetwork/lorawan-stack/runs/1684533672?check_suite_focus=true, but https://github.com/TheThingsNetwork/lorawan-stack/actions/runs/478857664 passes)

#### What do you want to see instead?
<!-- Please add some examples or mock-ups if applicable. -->

Tests should pass

#### Environment
<!--
Your environment: OS/Browser/Gateway/Device/...? Versions? IDs/EUIs?
Paste the output of "ttn-lw-cli version" or "ttn-lw-stack version" if applicable.
-->

- TTS 3.11
- Redis 5 (I am also getting errors locally with Redis 6, but CI seems to work just fine).

#### Can you do this yourself and submit a Pull Request?
<!-- You can also @mention experts if you need help with this. -->

No

Another CI failure here: https://github.com/TheThingsNetwork/lorawan-stack/pull/3586/checks?check_run_id=1684911932#step:14:143

<summary><details>

```
 ERROR Failed to generate class A downlink, skip class A downlink slot band_id=EU_863_870 dev_addr=42FFFFFF device_class=CLASS_A device_uid=process-downlink-test-app-id.process-downlink-test-dev-id downlink_type=data error=error:pkg/networkserver:no_downlink (no downlink to send) frequency_plan_id=EU_863_870 namespace=networkserver started_at=2021-01-12 00:18:00.3000015 +0000 UTC test_name=TestProcessDownlinkTask/Class_A/windows_open/1.0.3/RX1,RX2_available/no_MAC/generic_application_downlink/FCnt_too_low
 ERROR Failed to generate class A downlink, skip class A downlink slot band_id=EU_863_870 dev_addr=42FFFFFF device_class=CLASS_A device_uid=process-downlink-test-app-id.process-downlink-test-dev-id downlink_type=data error=error:pkg/networkserver:no_downlink (no downlink to send) frequency_plan_id=EU_863_870 namespace=networkserver started_at=2021-01-12 00:18:00.3000015 +0000 UTC test_name=TestProcessDownlinkTask/Class_A/windows_open/1.0.3/RX1,RX2_available/no_MAC/generic_application_downlink/application_downlink_exceeds_length_limit
--- FAIL: TestProcessDownlinkTask (0.49s)
    --- FAIL: TestProcessDownlinkTask/join-accept/windows_open/RX1,RX2_available/active_session/EU868 (0.02s)
        networkserver_util_internal_test.go:1978: events.Publish call missed: &{ctx:0xc0009a5590 innerEvent:{Name:ns.up.join.accept.forward Time:{wall:692578894 ext:63746007475 loc:<nil>} Identifiers:[0xc000499980] Data:<nil> CorrelationIDs:[] Origin:fv-az184-39 Context:map[] Visibility:0xc0004998c0 Authentication:<nil> RemoteIP: UserAgent: UniqueID:01EVSZ9KFCGJEHMPXWJXBYM4TP XXX_NoUnkeyedLiteral:{} XXX_sizecache:0} data:0xc001289280 caller:}
FAIL
coverage: 37.9% of statements
FAIL	go.thethings.network/lorawan-stack/v3/pkg/networkserver	0.587s
```

</details></summary>

More:

```
 ERROR Failed to generate class A downlink, skip class A downlink slot band_id=EU_863_870 dev_addr=42FFFFFF device_class=CLASS_A device_uid=process-downlink-test-app-id.process-downlink-test-dev-id@foo-tenant downlink_type=data error=error:pkg/networkserver:no_downlink (no downlink to send) frequency_plan_id=EU_863_870 namespace=networkserver started_at=2021-01-25 09:13:01.6600015 +0000 UTC tenant_id=foo-tenant test_name=TestProcessDownlinkTask/Class_A/windows_open/1.0.3/RX1,RX2_available/no_MAC/generic_application_downlink/FCnt_too_low
 ERROR Failed to generate class A downlink, skip class A downlink slot band_id=EU_863_870 dev_addr=42FFFFFF device_class=CLASS_A device_uid=process-downlink-test-app-id.process-downlink-test-dev-id@foo-tenant downlink_type=data error=error:pkg/networkserver:no_downlink (no downlink to send) frequency_plan_id=EU_863_870 namespace=networkserver started_at=2021-01-25 09:13:01.6600015 +0000 UTC tenant_id=foo-tenant test_name=TestProcessDownlinkTask/Class_A/windows_open/1.0.3/RX1,RX2_available/no_MAC/generic_application_downlink/application_downlink_exceeds_length_limit
--- FAIL: TestProcessDownlinkTask (0.36s)
    --- FAIL: TestProcessDownlinkTask/join-accept/windows_open/RX1,RX2_available/active_session/EU868 (0.02s)
        networkserver_util_internal_test.go:1984: events.Publish call missed: &{ctx:0xc000a62e40 innerEvent:{Name:ns.up.join.accept.forward Time:{wall:139388693 ext:63747162782 loc:<nil>} Identifiers:[0xc000bdbfa0] Data:<nil> CorrelationIDs:[] Origin:fv-az76-674 Context:map[] Visibility:0xc000bdbf80 Authentication:<nil> RemoteIP: UserAgent: UniqueID:01EWWD2RDVMJ7G7CKP901QE9G5 XXX_NoUnkeyedLiteral:{} XXX_sizecache:0} data:0xc000338f00 caller:}
FAIL
coverage: 37.9% of statements
FAIL	go.thethings.network/lorawan-stack/v3/pkg/networkserver	0.455s
```

--- a/pkg/networkserver/downlink_test.go
+++ b/pkg/networkserver/downlink_test.go
@@ -48,23 +48,17 @@ func TestProcessDownlinkTask(t *testing.T) {
 	appID := ttnpb.ApplicationIdentifiers{ApplicationID: appIDString}
 	const devID = "process-downlink-test-dev-id"
 
-	devAddr := types.DevAddr{0x42, 0xff, 0xff, 0xff}
-
-	fNwkSIntKey := types.AES128Key{0x42, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
-	nwkSEncKey := types.AES128Key{0x42, 0x42, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
-	sNwkSIntKey := types.AES128Key{0x42, 0x42, 0x42, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
-
 	joinAcceptBytes := append([]byte{0b001_000_00}, bytes.Repeat([]byte{0x42}, 32)...)
 
 	sessionKeys := &ttnpb.SessionKeys{
 		FNwkSIntKey: &ttnpb.KeyEnvelope{
-			Key: &fNwkSIntKey,
+			Key: &test.DefaultFNwkSIntKey,
 		},
 		NwkSEncKey: &ttnpb.KeyEnvelope{
-			Key: &nwkSEncKey,
+			Key: &test.DefaultNwkSEncKey,
 		},
 		SNwkSIntKey: &ttnpb.KeyEnvelope{
-			Key: &sNwkSIntKey,
+			Key: &test.DefaultSNwkSIntKey,
 		},
 		SessionKeyID: []byte{0x11, 0x22, 0x33, 0x44},
 	}
@@ -178,6 +172,16 @@ func TestProcessDownlinkTask(t *testing.T) {
 		}
 	}
 
+	attemptEventEqual := test.MakeEventEqual(test.EventEqualConfig{
+		Identifiers:    true,
+		CorrelationIDs: true,
+		Origin:         true,
+		Context:        true,
+		Visibility:     true,
+		Authentication: true,
+		RemoteIP:       true,
+		UserAgent:      true,
+	})
 	makeAssertReceiveScheduleFailAttemptEvents := func(attempt, fail events.Builder) func(context.Context, TestEnvironment, *ttnpb.DownlinkMessage, ttnpb.EndDeviceIdentifiers, error, uint) bool {
 		return func(ctx context.Context, env TestEnvironment, down *ttnpb.DownlinkMessage, ids ttnpb.EndDeviceIdentifiers, err error, n uint) bool {
 			_, a := test.MustNewTFromContext(ctx)
@@ -185,16 +189,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 			evIDOpt := events.WithIdentifiers(ids)
 			for i := uint(0); i < n; i++ {
 				if !test.AllTrue(
-					a.So(env.Events, should.ReceiveEventFunc, test.MakeEventEqual(test.EventEqualConfig{
-						Identifiers:    true,
-						CorrelationIDs: true,
-						Origin:         true,
-						Context:        true,
-						Visibility:     true,
-						Authentication: true,
-						RemoteIP:       true,
-						UserAgent:      true,
-					}),
+					a.So(env.Events, should.ReceiveEventFunc, attemptEventEqual,
 						attempt.With(events.WithData(down)).New(ctx, evIDOpt),
 					),
 					a.So(env.Events, should.ReceiveEventFunc, makeFailEventEqual(t),
@@ -212,16 +207,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 			ctx = events.ContextWithCorrelationID(ctx, down.CorrelationIDs...)
 			evIDOpt := events.WithIdentifiers(ids)
 			return test.AllTrue(
-				a.So(env.Events, should.ReceiveEventFunc, test.MakeEventEqual(test.EventEqualConfig{
-					Identifiers:    true,
-					CorrelationIDs: true,
-					Origin:         true,
-					Context:        true,
-					Visibility:     true,
-					Authentication: true,
-					RemoteIP:       true,
-					UserAgent:      true,
-				}),
+				a.So(env.Events, should.ReceiveEventFunc, attemptEventEqual,
 					attempt.With(events.WithData(down)).New(ctx, evIDOpt),
 				),
 				a.So(env.Events, should.ReceiveEventsResembling, events.Builders(append([]events.Builder{
@@ -255,7 +241,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 	_, ctx := test.New(t)
 	pingAt, ok := mac.NextPingSlotAt(ctx, &ttnpb.EndDevice{
 		Session: &ttnpb.Session{
-			DevAddr: devAddr,
+			DevAddr: test.DefaultDevAddr,
 		},
 		MACState: &ttnpb.MACState{
 			PingSlotPeriodicity: pingSlotPeriodicity,
@@ -343,7 +329,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 		Name                       string
 		CreateDevice               SetDeviceRequest
 		DeviceDiffs                []DeviceDiffFunc
-		ApplicationUplinkAssertion func(context.Context, *ttnpb.EndDevice, ...*ttnpb.ApplicationUp) bool
+		ApplicationUplinkAssertion func(context.Context, *ttnpb.EndDevice, ...*ttnpb.ApplicationUp) ([]events.Event, bool)
 		DownlinkAssertion          func(context.Context, TestEnvironment, *ttnpb.EndDevice) (*ttnpb.DownlinkMessage, time.Time, bool)
 		ErrorAssertion             func(*testing.T, error) bool
 	}{
@@ -358,7 +344,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -378,7 +364,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -393,7 +379,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								Matched:       true,
 								Confirmed:     true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								DataRateIndex: customCh.MinDataRateIndex,
 								Frequency:     customCh.UplinkFrequency,
@@ -410,7 +396,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						},
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 					},
@@ -432,7 +418,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -450,7 +436,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						},
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 					},
@@ -472,7 +458,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -487,7 +473,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								Matched:       true,
 								Confirmed:     true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								DataRateIndex: customCh.MinDataRateIndex,
 								Frequency:     customCh.UplinkFrequency,
@@ -514,7 +500,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -530,7 +516,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								Matched:       true,
 								Confirmed:     true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								DataRateIndex: customCh.MinDataRateIndex,
 								Frequency:     customCh.UplinkFrequency,
@@ -551,7 +537,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						StatusCountPeriodicity: &pbtypes.UInt32Value{Value: 0},
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 					},
@@ -574,7 +560,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -588,7 +574,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								DataRateIndex: customCh.MinDataRateIndex,
 								Frequency:     customCh.UplinkFrequency,
@@ -610,7 +596,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						StatusTimePeriodicity:  DurationPtr(0),
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 					},
@@ -633,7 +619,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_0_3_REV_A,
@@ -647,7 +633,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_0_3,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_0_3_REV_A].DataRates[customCh.MinDataRateIndex].Rate,
 								DataRateIndex: customCh.MinDataRateIndex,
 								Frequency:     customCh.UplinkFrequency,
@@ -669,7 +655,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						StatusTimePeriodicity:  DurationPtr(0),
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -704,9 +690,9 @@ func TestProcessDownlinkTask(t *testing.T) {
 			DeviceDiffs: []DeviceDiffFunc{
 				makeRemoveDownlinksDiff(2),
 			},
-			ApplicationUplinkAssertion: func(ctx context.Context, dev *ttnpb.EndDevice, ups ...*ttnpb.ApplicationUp) bool {
+			ApplicationUplinkAssertion: func(ctx context.Context, dev *ttnpb.EndDevice, ups ...*ttnpb.ApplicationUp) ([]events.Event, bool) {
 				_, a := test.MustNewTFromContext(ctx)
-				return a.So(ups, should.HaveLength, 1) &&
+				return nil, a.So(ups, should.HaveLength, 1) &&
 					a.So(ups[0], should.Resemble, &ttnpb.ApplicationUp{
 						EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
 						CorrelationIDs:       LastUplink(dev.MACState.RecentUplinks...).CorrelationIDs,
@@ -727,7 +713,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_0_3_REV_A,
@@ -741,7 +727,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_0_3,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_0_3_REV_A].DataRates[customCh.MinDataRateIndex].Rate,
 								DataRateIndex: customCh.MinDataRateIndex,
 								Frequency:     customCh.UplinkFrequency,
@@ -763,7 +749,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						StatusTimePeriodicity:  DurationPtr(0),
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -790,9 +776,9 @@ func TestProcessDownlinkTask(t *testing.T) {
 			DeviceDiffs: []DeviceDiffFunc{
 				makeRemoveDownlinksDiff(1),
 			},
-			ApplicationUplinkAssertion: func(ctx context.Context, dev *ttnpb.EndDevice, ups ...*ttnpb.ApplicationUp) bool {
+			ApplicationUplinkAssertion: func(ctx context.Context, dev *ttnpb.EndDevice, ups ...*ttnpb.ApplicationUp) ([]events.Event, bool) {
 				_, a := test.MustNewTFromContext(ctx)
-				return a.So(ups, should.HaveLength, 1) &&
+				return nil, a.So(ups, should.HaveLength, 1) &&
 					a.So(ups[0], should.Resemble, &ttnpb.ApplicationUp{
 						EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
 						CorrelationIDs:       append(LastUplink(dev.MACState.RecentUplinks...).CorrelationIDs, dev.Session.QueuedApplicationDownlinks[0].CorrelationIDs...),
@@ -813,7 +799,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -836,7 +822,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								DataRateIndex: customCh.MinDataRateIndex,
 								Frequency:     customCh.UplinkFrequency,
@@ -854,7 +840,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						RxWindowsAvailable: true,
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -892,7 +878,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_0_0110,
 							/*** FCnt ***/
@@ -901,8 +887,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/** FOpts **/
 						b = append(b, test.Must(crypto.EncryptDownlink(
-							nwkSEncKey,
-							devAddr,
+							test.DefaultNwkSEncKey,
+							test.DefaultDevAddr,
 							0x24,
 							[]byte{
 								/* ResetConf */
@@ -923,8 +909,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x42,
 							b,
@@ -973,7 +959,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -996,7 +982,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								DataRateIndex: customCh.MinDataRateIndex,
 								Frequency:     customCh.UplinkFrequency,
@@ -1014,7 +1000,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						RxWindowsAvailable: true,
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -1052,7 +1038,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_0_0110,
 							/*** FCnt ***/
@@ -1061,8 +1047,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/** FOpts **/
 						b = append(b, test.Must(crypto.EncryptDownlink(
-							nwkSEncKey,
-							devAddr,
+							test.DefaultNwkSEncKey,
+							test.DefaultDevAddr,
 							0x24,
 							[]byte{
 								/* ResetConf */
@@ -1083,8 +1069,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x42,
 							b,
@@ -1143,7 +1129,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -1166,7 +1152,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								DataRateIndex: customCh.MinDataRateIndex,
 								Frequency:     customCh.UplinkFrequency,
@@ -1184,7 +1170,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						RxWindowsAvailable: true,
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -1222,7 +1208,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_1_0110,
 							/*** FCnt ***/
@@ -1231,8 +1217,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/** FOpts **/
 						b = append(b, test.Must(crypto.EncryptDownlink(
-							nwkSEncKey,
-							devAddr,
+							test.DefaultNwkSEncKey,
+							test.DefaultDevAddr,
 							0x25,
 							[]byte{
 								/* ResetConf */
@@ -1247,8 +1233,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x25,
 							b,
@@ -1307,7 +1293,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -1330,7 +1316,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[ttnpb.DATA_RATE_6].Rate,
 								DataRateIndex: ttnpb.DATA_RATE_6,
 								Frequency:     customCh.UplinkFrequency,
@@ -1348,7 +1334,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						RxWindowsAvailable: true,
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -1386,7 +1372,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_0_0110,
 							/*** FCnt ***/
@@ -1395,8 +1381,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/** FOpts **/
 						b = append(b, test.Must(crypto.EncryptDownlink(
-							nwkSEncKey,
-							devAddr,
+							test.DefaultNwkSEncKey,
+							test.DefaultDevAddr,
 							0x24,
 							[]byte{
 								/* ResetConf */
@@ -1417,8 +1403,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x42,
 							b,
@@ -1474,7 +1460,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -1492,7 +1478,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								Frequency:     customCh.UplinkFrequency,
 								ChannelIndex:  customChIdx,
@@ -1511,7 +1497,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						},
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -1558,7 +1544,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_1_0000,
 							/*** FCnt ***/
@@ -1573,8 +1559,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x42,
 							b,
@@ -1623,7 +1609,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -1649,7 +1635,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								Frequency:     customCh.UplinkFrequency,
 								ChannelIndex:  customChIdx,
@@ -1666,7 +1652,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						RxWindowsAvailable: true,
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -1705,7 +1691,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_0_0110,
 							/*** FCnt ***/
@@ -1714,8 +1700,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/** FOpts **/
 						b = append(b, test.Must(crypto.EncryptDownlink(
-							nwkSEncKey,
-							devAddr,
+							test.DefaultNwkSEncKey,
+							test.DefaultDevAddr,
 							0x24,
 							[]byte{
 								/* ResetConf */
@@ -1736,8 +1722,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x42,
 							b,
@@ -1796,7 +1782,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -1822,7 +1808,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								Frequency:     customCh.UplinkFrequency,
 								ChannelIndex:  customChIdx,
@@ -1839,7 +1825,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						RxWindowsAvailable: true,
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -1877,7 +1863,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_0_0000,
 							/*** FCnt ***/
@@ -1892,8 +1878,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x42,
 							b,
@@ -1942,7 +1928,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:         test.EUFrequencyPlanID,
 					LastDevStatusReceivedAt: TimePtr(now),
@@ -1961,7 +1947,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								Frequency:     customCh.UplinkFrequency,
 								ChannelIndex:  customChIdx,
@@ -1978,7 +1964,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						RxWindowsAvailable: true,
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -2020,7 +2006,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_0_0000,
 							/*** FCnt ***/
@@ -2035,8 +2021,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x42,
 							b,
@@ -2086,7 +2072,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -2103,7 +2089,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								Frequency:     customCh.UplinkFrequency,
 								ChannelIndex:  customChIdx,
@@ -2119,7 +2105,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						},
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -2160,7 +2146,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_0_0000,
 							/*** FCnt ***/
@@ -2175,8 +2161,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x42,
 							b,
@@ -2233,9 +2219,9 @@ func TestProcessDownlinkTask(t *testing.T) {
 				removeMACQueueDiff,
 				makeRemoveDownlinksDiff(1),
 			},
-			ApplicationUplinkAssertion: func(ctx context.Context, dev *ttnpb.EndDevice, ups ...*ttnpb.ApplicationUp) bool {
+			ApplicationUplinkAssertion: func(ctx context.Context, dev *ttnpb.EndDevice, ups ...*ttnpb.ApplicationUp) ([]events.Event, bool) {
 				_, a := test.MustNewTFromContext(ctx)
-				return a.So(ups, should.HaveLength, 1) &&
+				return nil, a.So(ups, should.HaveLength, 1) &&
 					a.So(ups[0], should.Resemble, &ttnpb.ApplicationUp{
 						EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
 						CorrelationIDs:       dev.Session.QueuedApplicationDownlinks[0].CorrelationIDs,
@@ -2256,7 +2242,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -2273,7 +2259,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								Frequency:     customCh.UplinkFrequency,
 								ChannelIndex:  customChIdx,
@@ -2289,7 +2275,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						},
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -2330,7 +2316,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 							/* MACPayload */
 							/** FHDR **/
 							/*** DevAddr ***/
-							devAddr[3], devAddr[2], devAddr[1], devAddr[0],
+							test.DefaultDevAddr[3], test.DefaultDevAddr[2], test.DefaultDevAddr[1], test.DefaultDevAddr[0],
 							/*** FCtrl ***/
 							0b1_0_0_0_0000,
 							/*** FCnt ***/
@@ -2345,8 +2331,8 @@ func TestProcessDownlinkTask(t *testing.T) {
 
 						/* MIC */
 						mic := test.Must(crypto.ComputeDownlinkMIC(
-							sNwkSIntKey,
-							devAddr,
+							test.DefaultSNwkSIntKey,
+							test.DefaultDevAddr,
 							0,
 							0x42,
 							b,
@@ -2406,7 +2392,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -2425,7 +2411,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								Frequency:     customCh.UplinkFrequency,
 								ChannelIndex:  customChIdx,
@@ -2442,7 +2428,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						RxWindowsAvailable: true,
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -2478,7 +2464,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 					EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
 						ApplicationIdentifiers: appID,
 						DeviceID:               devID,
-						DevAddr:                &devAddr,
+						DevAddr:                &test.DefaultDevAddr,
 					},
 					FrequencyPlanID:   test.EUFrequencyPlanID,
 					LoRaWANPHYVersion: ttnpb.PHY_V1_1_REV_B,
@@ -2497,7 +2483,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 								DecodePayload: true,
 								Matched:       true,
 								MACVersion:    ttnpb.MAC_V1_1,
-								DevAddr:       devAddr,
+								DevAddr:       test.DefaultDevAddr,
 								DataRate:      LoRaWANBands[band.EU_863_870][ttnpb.PHY_V1_1_REV_B].DataRates[customCh.MinDataRateIndex].Rate,
 								Frequency:     customCh.UplinkFrequency,
 								ChannelIndex:  customChIdx,
@@ -2514,7 +2500,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						RxWindowsAvailable: true,
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -2575,7 +2561,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						QueuedJoinAccept: &ttnpb.MACState_JoinAccept{
 							Keys:    *sessionKeys,
 							Payload: joinAcceptBytes,
-							DevAddr: devAddr,
+							DevAddr: test.DefaultDevAddr,
 						},
 						RxWindowsAvailable: true,
 						RecentUplinks: []*ttnpb.UplinkMessage{
@@ -2598,7 +2584,7 @@ func TestProcessDownlinkTask(t *testing.T) {
 						},
 					},
 					Session: &ttnpb.Session{
-						DevAddr:       devAddr,
+						DevAddr:       test.DefaultDevAddr,
 						LastNFCntDown: 0x24,
 						SessionKeys:   *sessionKeys,
 						QueuedApplicationDownlinks: []*ttnpb.ApplicationDownlink{
@@ -2676,27 +2662,51 @@ func TestProcessDownlinkTask(t *testing.T) {
 					return true
 				},
 			},
-			ApplicationUplinkAssertion: func(ctx context.Context, dev *ttnpb.EndDevice, ups ...*ttnpb.ApplicationUp) bool {
-				return assertions.New(test.MustTFromContext(ctx)).So(ups, should.Resemble, []*ttnpb.ApplicationUp{
+			ApplicationUplinkAssertion: func(ctx context.Context, dev *ttnpb.EndDevice, ups ...*ttnpb.ApplicationUp) ([]events.Event, bool) {
+				a := assertions.New(test.MustTFromContext(ctx))
+				ids := ttnpb.EndDeviceIdentifiers{
+					ApplicationIdentifiers: dev.ApplicationIdentifiers,
+					DeviceID:               dev.DeviceID,
+					DevEUI:                 dev.DevEUI,
+					JoinEUI:                dev.JoinEUI,
+					DevAddr:                &dev.PendingMACState.QueuedJoinAccept.DevAddr,
+				}
+				cids := LastUplink(dev.PendingMACState.RecentUplinks...).CorrelationIDs
+				recvAt := LastUplink(dev.PendingMACState.RecentUplinks...).ReceivedAt
+
+				ok := a.So(ups, should.Resemble, []*ttnpb.ApplicationUp{
 					{
-						EndDeviceIdentifiers: ttnpb.EndDeviceIdentifiers{
-							ApplicationIdentifiers: dev.ApplicationIdentifiers,
-							DeviceID:               dev.DeviceID,
-							DevEUI:                 dev.DevEUI,
-							JoinEUI:                dev.JoinEUI,
-							DevAddr:                &dev.PendingMACState.QueuedJoinAccept.DevAddr,
-						},
-						CorrelationIDs: LastUplink(dev.PendingMACState.RecentUplinks...).CorrelationIDs,
+						EndDeviceIdentifiers: ids,
+						CorrelationIDs:       cids,
 						Up: &ttnpb.ApplicationUp_JoinAccept{
 							JoinAccept: &ttnpb.ApplicationJoinAccept{
 								AppSKey:              dev.PendingMACState.QueuedJoinAccept.Keys.AppSKey,
 								InvalidatedDownlinks: dev.Session.QueuedApplicationDownlinks,
 								SessionKeyID:         dev.PendingMACState.QueuedJoinAccept.Keys.SessionKeyID,
-								ReceivedAt:           LastUplink(dev.PendingMACState.RecentUplinks...).ReceivedAt,
+								ReceivedAt:           recvAt,
 							},
 						},
 					},
 				})
+				if !ok {
+					return nil, false
+				}
+				return []events.Event{
+					EvtForwardJoinAccept.With(
+						events.WithIdentifiers(ids),
+						events.WithData(&ttnpb.ApplicationUp{
+							EndDeviceIdentifiers: ids,
+							CorrelationIDs:       cids,
+							Up: &ttnpb.ApplicationUp_JoinAccept{
+								JoinAccept: &ttnpb.ApplicationJoinAccept{
+									InvalidatedDownlinks: dev.Session.QueuedApplicationDownlinks,
+									SessionKeyID:         dev.PendingMACState.QueuedJoinAccept.Keys.SessionKeyID,
+									ReceivedAt:           recvAt,
+								},
+							},
+						}),
+					).New(ctx),
+				}, true
 			},
 		},
 	} {
@@ -2807,9 +2817,15 @@ func TestProcessDownlinkTask(t *testing.T) {
 					}
 				}
 				if tc.ApplicationUplinkAssertion != nil {
+					var evs []events.Event
 					a.So(env.AssertNsAsHandleUplink(ctx, appID, func(ctx context.Context, ups ...*ttnpb.ApplicationUp) bool {
-						return tc.ApplicationUplinkAssertion(ctx, created, ups...)
+						var ok bool
+						evs, ok = tc.ApplicationUplinkAssertion(ctx, created, ups...)
+						return ok
 					}, nil), should.BeTrue)
+					if len(evs) > 0 {
+						a.So(env.Events, should.ReceiveEventsResembling, evs)
+					}
 				}
 			},
 		})

--- a/pkg/networkserver/grpc_asns.go
+++ b/pkg/networkserver/grpc_asns.go
@@ -80,6 +80,7 @@ func (ns *NetworkServer) processApplicationUplinkTask(ctx context.Context) error
 					return err
 				}
 				for _, up := range ups {
+					ctx := events.ContextWithCorrelationID(ctx, up.CorrelationIDs...)
 					switch pld := up.Up.(type) {
 					case *ttnpb.ApplicationUp_UplinkMessage:
 						registerForwardDataUplink(ctx, pld.UplinkMessage)

--- a/pkg/util/test/assertions/events.go
+++ b/pkg/util/test/assertions/events.go
@@ -214,7 +214,7 @@ func ShouldReceiveEventsFunc(actual interface{}, expected ...interface{}) string
 	return success
 }
 
-// ShouldReceiveEventsFunc is like ShouldReceiveEventsFunc, but uses same resemblance function as ShouldResembleEvent.
+// ShouldReceiveEventsResembling is like ShouldReceiveEventsFunc, but uses same resemblance function as ShouldResembleEvent.
 func ShouldReceiveEventsResembling(actual interface{}, expected ...interface{}) string {
 	if len(expected) == 0 {
 		return fmt.Sprintf(needAtLeastValues, 1, len(expected))
@@ -225,7 +225,7 @@ func ShouldReceiveEventsResembling(actual interface{}, expected ...interface{}) 
 	}
 	for i, exp := range evs {
 		if s := ShouldReceiveEventResembling(actual, exp); s != success {
-			return fmt.Sprintf("Mismatch in event number %d: %s", i, s)
+			return fmt.Sprintf("Mismatch in event number %d:\n%s", i, s)
 		}
 	}
 	return success

--- a/pkg/util/test/assertions/resemble.go
+++ b/pkg/util/test/assertions/resemble.go
@@ -49,7 +49,7 @@ func ShouldResemble(actual interface{}, expected ...interface{}) (message string
 		return success
 	}
 
-	diff := pretty.Diff(expected[0], actual)
+	diff := pretty.Diff(actual, expected[0])
 	if len(diff) == 0 {
 		return message
 	}
@@ -59,6 +59,8 @@ func ShouldResemble(actual interface{}, expected ...interface{}) (message string
 	for _, d := range diff {
 		lines = append(lines, fmt.Sprintf("   %s", d))
 	}
+	lines = append(lines, fmt.Sprintf("Actual: %s", pretty.Sprint(actual)))
+	lines = append(lines, fmt.Sprintf("Expected: %s", pretty.Sprint(expected[0])))
 	return strings.Join(append(lines, lastLine(message)), "\n")
 }
 

